pkg/nfd-topology-gc/nfd-nrt-gc.go

```diff
@@ -0,0 +1,219 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nfdtopologygarbagecollector
+
+import (
+	"context"
+	"time"
+
+	topologyclientset "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/generated/clientset/versioned"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/node-feature-discovery/pkg/apihelper"
+)
+
+// Args are the command line arguments
+type Args struct {
+	GCPeriod time.Duration
+
+	Kubeconfig string
+}
+
+type NRTGarbageCollector interface {
+	Run() error
+	Stop()
+}
+
+type nrtGarbageCollector struct {
+	stopChan   chan struct{}
+	topoClient topologyclientset.Interface
+	gcPeriod   time.Duration
+	factory    informers.SharedInformerFactory
+}
+
+func New(args *Args) (NRTGarbageCollector, error) {
+
+	kubeconfig, err := apihelper.GetKubeconfig(args.Kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	stop := make(chan struct{})
+
+	return newNRTGarbageCollector(kubeconfig, stop, args.GCPeriod)
+}
+
+func newNRTGarbageCollector(config *restclient.Config, stop chan struct{}, gcPeriod time.Duration) (*nrtGarbageCollector, error) {
+	helper := apihelper.K8sHelpers{Kubeconfig: config}
+	cli, err := helper.GetTopologyClient()
+	if err != nil {
+		return nil, err
+	}
+
+	clientset := kubernetes.NewForConfigOrDie(config)
+	factory := informers.NewSharedInformerFactory(clientset, 5*time.Minute)
+
+	return &nrtGarbageCollector{
+		topoClient: cli,
+		stopChan:   stop,
+		gcPeriod:   gcPeriod,
+		factory:    factory,
+	}, nil
+}
+
+func (n *nrtGarbageCollector) deleteNodeHandler(object interface{}) {
+	// handle a case when we are starting up and need to clear stale NRT resources
+	obj := object
+	if deletedFinalStateUnknown, ok := object.(cache.DeletedFinalStateUnknown); ok {
+		klog.V(2).Infof("found stale NodeResourceTopology for node: %s ", object)
+		obj = deletedFinalStateUnknown.Obj
+	}
+
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		klog.Errorf("cannot convert %v to v1.Node", object)
+		return
+	}
+	if err := n.topoClient.TopologyV1alpha1().NodeResourceTopologies().Delete(context.TODO(), node.GetName(), metav1.DeleteOptions{}); err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(2).Infof("NodeResourceTopology for node %s not found, omitting deletion", node.GetName())
+			return
+		} else {
+			klog.Warningf("failed to delete NodeResourceTopology for node %s: %s", node.GetName(), err.Error())
+			return
+		}
+	}
+	klog.Infof("NodeResourceTopology for node %s has been deleted", node.GetName())
+}
+
+// populateNodeIndexer populates cache with NRTs so we know which one to delete on first synchronization
+func (n *nrtGarbageCollector) populateNodeIndexer(indexer cache.Indexer) error {
```

Comment:
> Because we want to remove obsolete objects during first synchronization of indexer. The indexer will list current objects, compare it with what it have in store and run `DeleteHandler` on objects that are not present in k8s. We don't have to list `Nodes` and we can entire rely on indexer/lister wrt to Nodes. It's just an optimization. If it's more clear I can just list Nodes and compare it with NRTs and remove obsolete objects without populating cache.

Thanks for the explanation. I for myself appreciate the effort and the optimization, but in this case, especially considering the GC binary goals, I'd go for the obvious, trivial solution, albeit less efficient.
In general, if the GC process becomes a bottleneck or consumes a nontrivial amount of resources, some horribly broken already happened,  so I'm not worried about performances just yet.

Reply:
I'm inclined to agree with @ffromani here. I'd also prefer keeping the GC simple, just for the sake of ease of future maintainability and readability. Especially the startup which is done only once (or very rarely) doing a small amount of more work shouldn't really be a problem.

Reply:
simplified.